pkg/network/node/healthcheck.go

```diff
@@ -0,0 +1,98 @@
+package node
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/origin/pkg/util/ovs/ovsclient"
+)
+
+const (
+	ovsDialTimeout         = 5 * time.Second
+	ovsHealthcheckInterval = 30 * time.Second
+	ovsRecoveryTimeout     = 10 * time.Second
+)
+
+// waitForOVS polls until the OVS server responds to a connection and an 'echo'
+// command.
+func waitForOVS(network, addr string) error {
+	return utilwait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		c, err := ovsclient.DialTimeout(network, addr, ovsDialTimeout)
+		if err != nil {
+			glog.V(2).Infof("waiting for OVS to start: %v", err)
+			return false, nil
+		}
+		defer c.Close()
+		if err := c.Ping(); err != nil {
+			glog.V(2).Infof("waiting for OVS to start, ping failed: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+// runOVSHealthCheck runs two background loops - one that waits for disconnection
+// from the OVS server and then checks healthFn, and one that periodically checks
+// healthFn. If healthFn returns false in either of these two cases while the OVS
+// server is responsive the node process will terminate.
+func runOVSHealthCheck(network, addr string, healthFn func() bool) {
+	// this loop holds an open socket connection to OVS until it times out, then
+	// checks for health
+	go utilwait.Until(func() {
+		c, err := ovsclient.DialTimeout(network, addr, ovsDialTimeout)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("SDN healthcheck unable to connect to OVS server: %v", err))
+			return
+		}
+		defer c.Close()
+
+		err = c.WaitForDisconnect()
+		utilruntime.HandleError(fmt.Errorf("SDN healthcheck disconnected from OVS server: %v", err))
```

Comment:
I don't know anything about the OVS raw protocol, but if it eventually times out idle connections then this might result in spurious errors in the logs.

Reply:
It's possible to configure OVS to do timeouts, but at least ootb on our deployed systems it does not.  I also have the 5s disconnect.  I think if in practice we see this error showing up we would increase the timeout on connections significantly and still be ok.  It's effectively a deadman switch (and it works really well at it).